### PR TITLE
fix: make package independant from wdio testrunner

### PR DESCRIPTION
**Pull request by DudaGod**
Adopt changes from https://github.com/webdriverio/expect-webdriverio/commit/5e8c2b4c5a32340e1d7570f5303e04553a6c97e7 to v3.

Fix issue - https://github.com/webdriverio/expect-webdriverio/issues/924

### CODE REVIEW

**Comment by christian-bromann** [comment]
@DudaGod thanks for putting up the PR.

One thing I forgot is that in `v7` people technically can still run sync mode which means that all `browser.call` have to stay in place. In order to make this backwards compatible we should add them back. You can get the browser instance using this helper method:

```ts
/**
 * traverse up the scope chain until browser element was reached
 */
export function getBrowserObject (elem: WebdriverIO.Element | WebdriverIO.Browser): WebdriverIO.Browser {
    const elemObject = elem as WebdriverIO.Element
    return (elemObject as WebdriverIO.Element).parent ? getBrowserObject(elemObject.parent) : elem as WebdriverIO.Browser
}
```

So you can pass in a `WebdriverIO.Element` or `WebdriverIO.ElementArray` to get the `browser` instance.

**Comment by DudaGod** [comment]
> One thing I forgot is that in v7 people technically can still run sync mode which means that all browser.call have to stay in place. In order to make this backwards compatible we should add them back. 

ok, I returned it

> You can get the browser instance using this helper method:

But why I nee this helper If inside `waitUntil` helper I can stop to use `browser.waitUntil` and use simple js with `while`?

Fixed.

**Comment by christian-bromann** [comment]
> But why I need this helper If inside `waitUntil` helper I can stop to use `browser.waitUntil` and use simple js with `while`?

There are some other places where the global is used, e.g.: https://github.com/DudaGod/expect-webdriverio/blob/fix_in_stanalone_script.v3/src/matchers/element/toBeClickable.ts#L8

File: src/utils.ts [hunk 1]
```diff
@@ -209,6 +222,20 @@ function aliasFn(
...
 
+/**
+ * traverse up the scope chain until browser element was reached
+ */
+function getBrowserObject (elem: WebdriverIO.Element | WebdriverIO.Browser): WebdriverIO.Browser {
+    const elemObject = elem as WebdriverIO.Element
+    return (elemObject as WebdriverIO.Element).parent ? getBrowserObject(elemObject.parent) : elem as WebdriverIO.Browser
+}
+
+async function getFirstElement (received: WdioElementMaybePromise): Promise<WebdriverIO.Element> {
```
Comment (DudaGod) [inline src/utils.ts:233]: Added helper to not duplicate logic of get first element instance to get a browser instance

File: src/matchers/mock/toBeRequestedTimes.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import { waitUntil, enhanceError, compareNumbers } from '../../utils'
 import { runExpect } from '../../util/expectAdapter'
 import { numberError } from '../../util/formatMessage'
 
-export function toBeRequestedTimesFn(received: Mock, expected: number| ExpectWebdriverIO.NumberOptions = {}, options: ExpectWebdriverIO.StringOptions = {}): any {
+export function toBeRequestedTimesFn(received: Mock & {browser: WebdriverIO.Browser}, expected: number| ExpectWebdriverIO.NumberOptions = {}, options: ExpectWebdriverIO.StringOptions = {}): any {
```
Comment (DudaGod) [inline src/matchers/mock/toBeRequestedTimes.ts:7]: In mock functions I can't get browser instance. But I see that `Interception` type from which `Mock` is inherited has `browser` field - https://github.com/webdriverio/webdriverio/blob/428a9d729ae6231968a60908732fa3f607d195e9/packages/webdriverio/src/utils/interception/index.ts#L30

So I add this small hack which is working. If you have any ideas how to make it better, then I'm ready to fix it.

**Comment by DudaGod** [comment]
> There are some other places where the global is used, e.g.: https://github.com/DudaGod/expect-webdriverio/blob/fix_in_stanalone_script.v3/src/matchers/element/toBeClickable.ts#L8

Oh, yeah. You are right. Fixed.

File: test/__fixtures__/wdio.js [hunk 3]
```diff
@@ -39,7 +49,7 @@ function $(selector, ...args) {
 
 function $$(selector, ...args) {
     const length = this._length || 2
-    let els = Array(length).map((_, index) => {
+    let els = Array.from(Array(length), (_, index) => {
```
Comment (DudaGod) [inline test/__fixtures__/wdio.js:52]: An empty array was created here so I fixed it

File: test/__fixtures__/wdio.js [hunk 4]
```diff
@@ -27,7 +36,8 @@ const element = {
     isFocused: beFn,
     isEnabled: beFn,
     getProperty: getPropertyFn,
-    getText: getTextFn
+    getText: getTextFn,
+    parent: browser
```
Comment (DudaGod) [inline test/__fixtures__/wdio.js:40]: In order to method `getBrowserObject` works correctly

File: test/matchers/mock/toBeRequested.test.ts [hunk 5]
```diff
@@ -11,6 +11,9 @@ class TestMock {
     get calls () {
         return this._calls
     }
+    get browser () {
```
Comment (DudaGod) [inline test/matchers/mock/toBeRequested.test.ts:14]: Browser instance exists in `Mock` instance, so I added it here

**Comment by DudaGod** [comment]
Tests are fixed

**Comment by Baskercarrer** [comment]
Hi guys,

is it possible to release this PR?

**Comment by christian-bromann** [comment]
> is it possible to release this PR?

What made you think we would not do that?

**Comment by DudaGod** [comment]
@christian-bromann can you review it please?

**Review by christian-bromann — APPROVED** [review]
LGTM 👍

**Pull request merged**